app/domain/authentication/authn_oidc/authentication_service.rb

```diff
@@ -0,0 +1,101 @@
+require 'uri'
+require 'openid_connect'
+
+module Authentication
+  module AuthnOidc
+    class AuthenticationService
+      attr_reader :service_id
+      attr_reader :conjur_account
+
+      # Constructs AuthenticationService from the <service-id>, which is typically something like
+      # conjur/authn-oidc/<service-id>.
+      def initialize(service_id, conjur_account)
+        @service_id = service_id
+        @conjur_account = conjur_account
+
+        validate_variable_exists("client-id")
+        validate_variable_exists("client-secret")
+        validate_variable_exists("provider-uri")
+      end
+
+      # Retrieves an id token from the OpenIDConnect Provider and returns it decoded
+      def get_user_details (request_body)
+        request_body = URI.decode_www_form(request_body)
+        @redirect_uri = request_body.assoc('redirect_uri').last
+        authorization_code = request_body.assoc('code').last
+
+        client = get_client
+        client.authorization_code = authorization_code
+        access_token = client.access_token!
+
+        client.host = URI.parse(provider_uri).host
+        id_token = decode_id_token(access_token.id_token)
+        user_info = access_token.userinfo!
+
+        return id_token, user_info
+      rescue => e
+        raise OIDCAuthenticationError.new(e)
+      end
+
+      def client_id
+        Resource["#{conjur_account}:variable:#{service_id}/client-id"].secret.value
+      end
+
+      def issuer
+        discover.issuer
+      end
+
+      private
+
+      def validate_variable_exists (variableName)
+        resource = Resource["#{conjur_account}:variable:#{service_id}/#{variableName}"]
+        if resource.nil? || resource.secret.nil?
+          raise OIDCConfigurationError, "Variable [#{service_id}/#{variableName}] not found in Conjur"
+        end
+      end
+
+      def client_secret
+        Resource["#{conjur_account}:variable:#{service_id}/client-secret"].secret.value
+      end
+
+      def provider_uri
+        Resource["#{conjur_account}:variable:#{service_id}/provider-uri"].secret.value
+      end
+
+      def redirect_uri
+        @redirect_uri
+      end
+
+      def discover
+        # TODO: should not run in production
+        disable_ssl_verification
```

Comment:
I'm confused here.  Is this test only code?

Reply:
no. we'll add the ssl verification later

Reply:
We have an issue with our OpenID 3rd party lib, which doesn’t support self sign certificate, and prevent us to communicate with keycloak (the OpenID provider in our dev&test environment ),
We aware to this issue and we planned to handle this in the new design of the 3 APIs ( after this PR).